### PR TITLE
feat: add kotlin support for model generation

### LINKED ISSUE #432: Add Kotlin support for model generation
#### Reason/Context
As of Modelina version `1.0.0-next.41` Kotlin is supported as a generator, which the CLI also need to support.

To solve this issue you must do the following:

1. Update Modelina to the new version.
2. Add Kotlin to the list of supported languages: https://github.com/asyncapi/cli/blob/c2cf56d5bfbc8ab424ad89dfc2eb31b042cbca35/src/commands/generate/models.ts#L6
3. Add kotlin to the list of languages that needs `packageName` argument: https://github.com/asyncapi/cli/blob/c2cf56d5bfbc8ab424ad89dfc2eb31b042cbca35/src/commands/generate/models.ts#L67
4. Import the `KotlinFileGenerator`: https://github.com/asyncapi/cli/blob/c2cf56d5bfbc8ab424ad89dfc2eb31b042cbca35/src/commands/generate/models.ts#L1
5. Add the following to [the switch case](https://github.com/asyncapi/cli/blob/c2cf56d5bfbc8ab424ad89dfc2eb31b042cbca35/src/commands/generate/models.ts#L102):
```ts
    case Languages.kotlin:
      if (packageName === undefined) {
        throw new Error('In order to generate models to Kotlin, we need to know which package they are under. Add `--packageName=PACKAGENAME` to set the desired package name.');
      }
      fileGenerator = new KotlinFileGenerator();
      fileOptions = {
        packageName
      };
```
6. Add the [following tests for Kotlin](https://github.com/asyncapi/cli/blob/master/test/commands/generate/models.test.ts#L141):
```ts
  describe('for Kotlin', () => {  
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'kotlin', './test/specification.yml', `-o=${ path.resolve(outputDir, './kotlin')}`, '--packageName', 'test.package'])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'kotlin', './test/specification.yml', `-o=${ path.resolve(outputDir, './kotlin')}`])
      .it('fails when no package defined', (ctx, done) => {
        expect(ctx.stderr).toEqual('Error: In order to generate models to Kotlin, we need to know which package they are under. Add `--packageName=PACKAGENAME` to set the desired package name.\n');
        expect(ctx.stdout).toEqual('');
        done();
      });
  });
```
7. Add the Kotlin command to the main readme: https://github.com/asyncapi/cli/blob/master/README.md#usage

That is all that is needed to support Kotlin in the CLI.

### PR DESCRIPTION
**Pull request by sambhavgupta0705**
**Description**

- ... added kotlin support for model generation.

**Related issue(s)**

Resolves #432

### CODE REVIEW

**Comment by sambhavgupta0705** [comment]
cc @jonaslagoni @derberg

**Review by jonaslagoni — DISMISSED** [review]
LGTM 👍

**Comment by sambhavgupta0705** [comment]
@derberg PTAL

File: src/commands/generate/models.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { CSharpFileGenerator, JavaFileGenerator, JavaScriptFileGenerator, TypeScriptFileGenerator, GoFileGenerator, Logger, DartFileGenerator, PythonFileGenerator, RustFileGenerator } from '@asyncapi/modelina';
+import { CSharpFileGenerator, JavaFileGenerator, JavaScriptFileGenerator, TypeScriptFileGenerator, GoFileGenerator, Logger, DartFileGenerator, PythonFileGenerator, RustFileGenerator ,KotlinFileGenerator} from '@asyncapi/modelina';
```
Comment (derberg) [inline src/commands/generate/models.ts:1]:
```suggestion
import { CSharpFileGenerator, JavaFileGenerator, JavaScriptFileGenerator, TypeScriptFileGenerator, GoFileGenerator, Logger, DartFileGenerator, PythonFileGenerator, RustFileGenerator, KotlinFileGenerator} from '@asyncapi/modelina';
```

File: src/commands/generate/models.ts [hunk 2]
```diff
@@ -64,7 +65,7 @@ export default class Models extends Command {
      * Go and Java specific package name to use for the generated models
      */
     packageName: Flags.string({
-      description: 'Go and Java specific, define the package to use for the generated models. This is required when language is `go` or `java`.',
+      description: 'Go ,Java and Kotlin specific, define the package to use for the generated models. This is required when language is `go`, `java` or `kotlin`.',
```
Comment (derberg) [inline src/commands/generate/models.ts:68]:
```suggestion
      description: 'Go, Java and Kotlin specific, define the package to use for the generated models. This is required when language is `go`, `java` or `kotlin`.',
```

**Review by derberg — COMMENTED** [review]
2 tiny suggestions

File: package.json [hunk 3]
```diff
@@ -11,7 +11,7 @@
     "@asyncapi/converter": "^1.1.0",
     "@asyncapi/diff": "^0.4.1",
     "@asyncapi/generator": "^1.9.12",
-    "@asyncapi/modelina": "^1.0.0-next.40",
+    "@asyncapi/modelina": "^1.0.0-next.43",
```
Comment (derberg) [inline package.json:14]: please use latest modelina, main release not release candidate 

it was released few days ago

**Comment by sambhavgupta0705** [comment]
@derberg made all the suggested changes

File: package.json (hunk 3 above)
Comment (derberg) [reply package.json]: https://github.com/asyncapi/modelina/releases/tag/v1.1.1

Comment (derberg) [reply package.json]: ok, I see you changed it but there are anyway conflicts you need to solve as there were some releases in CLI yesterday

Comment (sambhavgupta0705) [reply package.json]: Ok I will look into this one

**Comment by derberg** [comment]
@sambhavgupta0705 sorry but due to recent development again many changes were pushed to `master` that are related to dependencies and you have merge conflicts again 😓 

also there is small change in docs. The change from readme you need to move to proper markdown document in `docs` folder (that you will see after merging `master` to your branch)

**Review by derberg — DISMISSED** [review]
🚀

**Review by derberg — APPROVED** [review]

**Review by derberg — COMMENTED** [review]
@jonaslagoni 

hey man, can you help with tests here. @sambhavgupta0705 did all good here, I don't see anything he changed that could break tests.

so, I checked locally and it really does not work:
```
./bin/run generate models typescript http://bit.ly/asyncapi

TypeError: Cannot read property 'constructor' of undefined
```

any ideas what that could be?

File: test/commands/generate/models.test.ts [hunk 4]
```diff
@@ -161,6 +161,29 @@ describe('models', () => {
       });
   });
 
+  describe('for Kotlin', () => {  
+    test
+      .stderr()
+      .stdout()
+      .command([...generalOptions, 'kotlin', './test/specification.yml', `-o=${ path.resolve(outputDir, './kotlin')}`, '--packageName', 'test.package'])
```
Comment (jonaslagoni) [inline test/commands/generate/models.test.ts:168]:
```suggestion
      .command([...generalOptions, 'kotlin', './test/specification.yml', `-o=${ path.resolve(outputDir, './kotlin')}`, '--packageName', 'asyncapi.models'])
```

As of version 1.0.0-next.42 `test.package` is marked as an incorrect package name: https://github.com/asyncapi/modelina/pull/1084

You need to adapt the Java test as well.

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
The only error I see is the test for package name 🤔

**Comment by jonaslagoni** [comment]
But this error:
```
TypeError: Cannot read property 'constructor' of undefined
```

Is weird.

**Comment by sambhavgupta0705** [comment]
Okay @jonaslagoni will look into it.

**Comment by jonaslagoni** [comment]
It seems like it's an issue in Modelina 🧐 Ill get back to you.

**Comment by jonaslagoni** [comment]
Just checked if It works in an independent library, and it did 🤷

**Comment by derberg** [comment]
@jonaslagoni interesting is that when I run generation for typescript, it fails but java and javascript it works like a charm.

When you checked in independent library, what language did you check?

**Comment by jonaslagoni** [comment]
TypeScript 🤨

**Comment by jonaslagoni** [comment]
I did recently have some issues with a dependency used in a TypeScript preset though... I wonder if that is causing problems here as well 🧐 https://github.com/sourcemeta/alterschema/pull/110

**Comment by derberg** [comment]
fun fact, when I remove all TS tests and fix java `test.package` test - they are passing

**Comment by jonaslagoni** [comment]
Without debugging capabilities as simple as getting a full stack trace, this becomes almost impossible 😄 

But yea, facts so far, the TypeScript generator works outside of the CLI repo, at least what I could reproduce, all other generators works besides the TypeScript one in the CLI, when you remove the typescript code.

Does that pretty much sum it up so far @derberg ? 🙂

**Comment by sambhavgupta0705** [comment]
Hey @jonaslagoni can you sum up that what changes do I need to make in the PR?

**Comment by jonaslagoni** [comment]
@sambhavgupta0705 that is still to be determined 😆 It's a weird case.

**Comment by sambhavgupta0705** [comment]
@jonaslagoni Lol ,okay then I will wait.

**Comment by derberg** [comment]
@sambhavgupta0705 for now you can fix the test that is not related to `constructor` error but `test.package` in Java test. Just rename `package` to something else, maybe `asyncapi`

**Comment by sambhavgupta0705** [comment]
ok @derberg will try doing that

**Comment by sambhavgupta0705** [comment]
Hey @derberg & @jonaslagoni , I am not able to understand how to solve errors manually ,can you please guide me a little

**Comment by jonaslagoni** [comment]
@sambhavgupta0705 I reached out to the community on Slack, as I, unfortunately, don't have the time to find the root cause. Not sure about you @derberg?

https://asyncapi.slack.com/archives/CQVJXFNQL/p1676402918618609

**Comment by derberg** [comment]
yeah, I'll try to delegate some time in next few days as I'm getting tired of this tests in CLI, they're flaky as fuck 😄 my finger got tired of clicking `rerun`

**Comment by peter-rr** [comment]
> ```
> ./bin/run generate models typescript http://bit.ly/asyncapi
> 
> TypeError: Cannot read property 'constructor' of undefined
> ```
> 
> any ideas what that could be?

Hey guys, I was getting the same weird error when running tests and have come up with a fix to solve the failing tests related to generation of models. I just had to add some default values for “TypeScript specific options” at `models.ts` to avoid `undefined` values which were causing the error. Do you want me to create a PR for this fix? @derberg @jonaslagoni

**Comment by jonaslagoni** [comment]
Please do @peter-rr! Thanks! 

Might even need some checks in Modelina to throw a proper error.

**Comment by derberg** [comment]
@sambhavgupta0705 please resolve conflicts and update to latest master. Thanks to https://github.com/asyncapi/cli/pull/458 your issues should be solve 🤞🏼

**Comment by sambhavgupta0705** [comment]
@derberg I will be able to work on this after this weak as my exams are going on.

**Comment by jonaslagoni** [comment]
@sambhavgupta0705 ping 😄

**Comment by sambhavgupta0705** [comment]
Yes @jonaslagoni will be resolving this super soon.

**Comment by sambhavgupta0705** [comment]
@jonaslagoni I made all the changes but still some tests are failing 
image

**Comment by sambhavgupta0705** [comment]
I also tried doing all this on a new branch with only necessary but still some test are failing

**Comment by sambhavgupta0705** [comment]
> @sambhavgupta0705 have you merged the current master branch into your branch?

yes @jonaslagoni merged all the changes 
image
This is the end result after running `npm run test`

**Comment by jonaslagoni** [comment]
@sambhavgupta0705 I have no idea why this is happening, there are so many variables that might cause this from a dependency point of view... I am updating Modelina to v1 in https://github.com/asyncapi/cli/pull/477 that you can then merge into your branch, and then it SHOULD work out of the box 🙃

**Comment by sambhavgupta0705** [comment]
> Try update your branch @sambhavgupta0705 🙂 

Ok Jonas,will do it soon

**Comment by sambhavgupta0705** [comment]
Closing this PR and made a new one with all the changes #480

**Pull request closed without merging**